add field and label selectors to authorization

liggitt opened the pull request:
Contains the work for [KEP 4601](https://github.com/kubernetes/enhancements/issues/4601)

Based on https://github.com/kubernetes/kubernetes/pull/125371 with additional commits for:
- improving existing CEL cost testing
- fixing existing CEL env construction happening in init() functions (which would not be informed by feature-gates or compatibility-version adjustments)
- Adding CEL support for authorization webhook matchConditions
- Adding CEL support for CRD/admission webhook authorizer library
- Adding node authorizer use of selector validation

Follow-ups:
* @jpbetz: improve library compatibility test registration (https://github.com/kubernetes/kubernetes/pull/125571#discussion_r1658894149)
* @jpbetz: improve cost unit tests to flag unhandled / defaulted costs (https://github.com/kubernetes/kubernetes/pull/125571#discussion_r1657860594)
* @jpbetz: TODO to untangle use of environment.DefaultCompatibilityVersion() from spots that run before compatibility version options are parsed
* https://github.com/kubernetes/kubernetes/issues/125775
* @liggitt - KEP updates:
  * add deletecollection verb (https://github.com/kubernetes/kubernetes/pull/125571#discussion_r1658830850)
  * separate feature gate for node authorizer bits, impact to users piggy-backing on kubelet credentials and doing broad reads, how to react (https://github.com/kubernetes/kubernetes/pull/125571#discussion_r1658940126)
  * *SAR parsing errors return indicator in status.evaluationError (https://github.com/kubernetes/kubernetes/pull/125571#discussion_r1658882844)
  * renamed Parse*Selector() to Get*Selector() in interface (https://github.com/kubernetes/kubernetes/pull/125571#discussion_r1658845772)

/priority important-soon

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
kube-apiserver: Added alpha features to allow API server authz to check the context of requests:
- The `AuthorizeWithSelectors` feature gate enables including field and label selector information from requests in webhook authorization calls.
- The `AuthorizeNodeWithSelectors` feature gate changes node authorizer behavior to limit requests from node API clients, so that each Node can only get / list / watch its own Node API object, and can also only get / list / watch Pod API objects bound to that node. Clients using kubelet credentials to read other nodes or unrelated pods must change their authentication credentials (recommended), adjust their usage, or obtain broader read access independent of the node authorizer.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/4601
```

/sig auth
/cc @enj @deads2k

deads2k committed:
add field and label selectors to authorization attributes

Co-authored-by: Jordan Liggitt <liggitt@google.com>

deads2k committed:
add subjectaccessreview field and label selectors

Co-authored-by: Jordan Liggitt <liggitt@google.com>

deads2k committed:
generate

deads2k committed:
Add CEL fieldSelector / labelSelector support to authorizer library

liggitt committed:
Add structured labelSelector / fieldSelector to authorization webhook match conditions

liggitt commented:
/retest

liggitt committed:
Adjust CEL cost calculation and versioning for authorization library

liggitt committed:
Improve CEL cost tests to catch unhandled estimates or types

liggitt committed:
Add selector authorization to the Node authorizer

liggitt committed:
Fixup lint warning

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/library/cost.go:
```diff
@@ -25,9 +26,28 @@ import (
 
+// panicOnUnknown makes cost estimate functions panic on unrecognized functions.
+// This is only set to true for unit tests.
+var panicOnUnknown = false
```
@jpbetz I got bit when implementing cost tests on the selector functions and was hitting unknown paths and the cost estimates were silently defaulting to 0 / 1, etc... I added this to make those unhandled paths loud in unit test mode ... thoughts on this?

jpbetz replied:
Something like this is needed.  Turning this on for all our existing cost tests would be great.

jpbetz replied:
It might be possible to come up with a test that ensures we try out every library function with this enabled. I'd be willing to take that on as a follow up.

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/library/cost.go:
```diff
@@ -185,6 +205,13 @@ func (l *CostEstimator) CallCost(function, overloadId string, args []ref.Val, re
 	case "sign", "asInteger", "isInteger", "asApproximateFloat", "isGreaterThan", "isLessThan", "compareTo", "add", "sub":
 		cost := uint64(1)
 		return &cost
+	case "getScheme", "getHostname", "getHost", "getPort", "getEscapedPath", "getQuery":
```
this is good example of unhandled paths which got missed and were hitting the default cost of 1 ... fortunately that is a reasonable cost for these, but we could easily have missed cases for expensive things

jpbetz replied:
+100

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/library/authz.go:
```diff
@@ -256,12 +275,26 @@ var authzLibraryDecls = map[string][]cel.FunctionOpt{
 		options = append(options, cel.Function(name, overloads...))
 	}
+	if a.version >= 1 {
+		// TODO(liggitt): only allow expressions with these to be persisted when AuthorizeWithSelectors feature is enabled?
```
@jpbetz do you have a recommendation here? I'm not sure how feature gate enablement interacts with the versioned library registration

jpbetz replied:
I think this will be the first time feature gate a library.

I was imagining that we'd control it at the environment level (e.g. cel/environment/base.go) rather than at the library level.

The environment configuration doesn't support this case quite right though. So we'll have to modify it.

In particular, we need to make it possible to introduce a new library version (we can do that), and enable it for `StoredExpressions` (we can also do that), but don't make it enable automatically in the next release for `NewExpressions` (we can't do that).

In this use case, I think we also need to be able to `Extend` the base environment for the validation of the new API CEL fields so that they can immediately use the library version for `NewExpressions` _if_ the feature gate is enabled.  Conditionally extending the base environment using a feature gate is possible. But extending an environment to enable a different version of a library that's already in the base environment is not something we've had to do before. So we'll have to add support for that.

Does that rough direction sound right? I can help with the details tomorrow if we're aligned on that part.

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/library/cost.go:
```diff
@@ -55,6 +55,18 @@ type CostEstimator struct {
+	costPerRequirement  = common.ListCreateBaseCost + common.StructCreateBaseCost
+)
+
+func selectorCostEstimate(selectorLength float64) uint64 {
```
@jpbetz can you sanity check this method of estimating cost for a selector based on string length?

jpbetz replied:
LGTM.

The `parseCost` and `maxRequirements` math looks sane and safe.  The object construction costs were originally intended to reflect the cost of constructing data literals within the CEL runtime. Using them here as a proxy measure for golang allocs should certainly bound the cost.

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/library/cost.go:
```diff
@@ -66,6 +78,13 @@ func (l *CostEstimator) CallCost(function, overloadId string, args []ref.Val, re
 		return &cost
+	case "fieldSelector", "labelSelector":
+		// field and label selector parse is a string parse into a structured set of requirements
+		if len(args) >= 2 {
```
@jpbetz it's a sharp edge that in CallCost the function args are arg[0]==receiver, arg[1+] are the args, and in EstimateCallCost, target is the receiver and args are the args ... any way we can make that more visible or clearer?

jpbetz replied:
We (CEL devs) can follow up on this one. I've opened https://github.com/kubernetes/kubernetes/issues/125775 to track.

liggitt commented on staging/src/k8s.io/apiserver/pkg/authorization/cel/compile.go:
```diff
@@ -203,6 +259,42 @@ func convertObjectToUnstructured(obj *authorizationv1.SubjectAccessReviewSpec) m
+
+		if includeFieldSelector && obj.ResourceAttributes.FieldSelector != nil {
+			if len(obj.ResourceAttributes.FieldSelector.Requirements) > 0 {
+				requirements := make([]map[string]interface{}, 0, len(obj.ResourceAttributes.FieldSelector.Requirements))
```
@jpbetz is there a better way to lazily adapt an existing instance to cel-expression-accessible ref.Vals? I really dislike reconstructing a deep struct like this

jpbetz replied:
Not yet :( CEL's direct support is only "primitive types in addition to lists, maps..." ([xref](https://github.com/google/cel-go/?tab=readme-ov-file#environment-setup)).  We added custom support for unstructured in k8s, and have done everything through that so far.  We have experience in SMD [wrapping go structs](https://github.com/kubernetes-sigs/structured-merge-diff/blob/master/value/valuereflect.go), which we could re-apply here.  I've been wanting to get that in place..

liggitt commented on pkg/registry/authorization/util/helpers.go:
```diff
@@ -35,6 +41,89 @@ func ResourceAttributesFrom(user user.Info, in authorizationapi.ResourceAttribut
+	if utilfeature.DefaultFeatureGate.Enabled(genericfeatures.AuthorizeWithSelectors) {
+		if in.LabelSelector != nil {
+			if len(in.LabelSelector.RawSelector) > 0 {
+				labelSelector, err := labels.Parse(in.LabelSelector.RawSelector)
```
we already have the status.evaluationError string field, which I think is perfect for communicating this back

liggitt replied:
TODO: update this detail in the KEP

liggitt commented on staging/src/k8s.io/apiserver/pkg/authorization/authorizer/interfaces.go:
```diff
@@ -146,6 +164,22 @@ func (a AttributesRecord) GetPath() string {
+	return a.FieldSelectorRequirements, a.FieldSelectorParsingErr
+}
+
+func (a AttributesRecord) ParseLabelSelector() (labels.Requirements, error) {
```
TODO: update this detail in the KEP

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/request/requestinfo.go:
```diff
@@ -238,11 +247,26 @@ func (r *RequestInfoFactory) NewRequestInfo(req *http.Request) (*RequestInfo, er
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(genericfeatures.AuthorizeWithSelectors) {
+		if requestInfo.Verb == "list" || requestInfo.Verb == "watch" {
```
TODO: update this detail in the KEP

liggitt commented on pkg/features/kube_features.go:
```diff
@@ -67,6 +67,14 @@ const (
+	//
+	// Make the Node authorizer use fine-grained selector authorization.
+	// Requires AuthorizeWithSelectors to be enabled.
+	AuthorizeNodeWithSelectors featuregate.Feature = "AuthorizeNodeWithSelectors"
```
TODO: update this detail in the KEP

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/library/cost.go:
```diff
@@ -227,6 +246,15 @@ func (l *CostEstimator) EstimateCallCost(function, overloadId string, target *ch
+		// field and label selector parse is a string parse into a structured set of requirements
+		if len(args) == 1 {
+			selectorSize := l.sizeEstimate(args[0])
+			return &checker.CallEstimate{CostEstimate: checker.CostEstimate{
```
switched this to propagating the estimate into selectorCostEstimate and have it use CostEstimate methods to add / multiply

liggitt commented:
Addressed all comments in `!fixup` commits, except the TODO about informing CEL env construction with feature gate enablement. Will look at that in a bit (and add some integration tests for the CEL bits as part of that).

liggitt committed:
Move CEL env initialization out of package init()

This ensures compatibility version and feature gates can be initialized
before cached CEL environments are created.

liggitt commented:
all comments addressed, feature-gating added to the AuthzSelectors library (verified attempts to use the new library functions with the feature disabled are rejected by validation, attempts to use the new library functions with the feature enabled are allowed)

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/environment/base.go:
```diff
@@ -149,13 +152,34 @@ var baseOptsWithoutStrictCost = []VersionedOptions{
-		// TODO(liggitt): limit this to feature enablement
+		FeatureEnabled: func() bool {
+			enabled := utilfeature.DefaultFeatureGate.Enabled(genericfeatures.AuthorizeWithSelectors)
+			authzSelectorsLibraryInit.Do(func() {
+				// Record the first time feature enablement was checked for this library.
+				// This is checked from integration tests to ensure no cached cel envs
+				// are constructed before feature enablement is effectively set.
+				authzSelectorsLibraryEnabled.Store(enabled)
+				// Uncomment to debug where the first initialization is coming from if needed.
+				// debug.PrintStack()
+			})
```
this is solely for the benefit of the integration tests making sure feature gates get parsed before AuthzSelector ever has to check for enablement

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/environment/environment.go:
```diff
@@ -175,7 +175,15 @@ type VersionedOptions struct {
+	// The StoredExpressions environment ignores this function.
+	//
+	// Optional.
+	FeatureEnabled func() bool
```
@jpbetz, this was what I came up with to make feature-gate-informed VersionedOptions.

StoredExpressions assumes gated libraries are always available and *only* goes off of IntroducedVersion / RemovedVersion, since stuff in storage could have gotten in when a feature gate was enabled.

Making it override IntroducedVersion / RemovedVersion for NewExpressions allows you to opt into an alpha feature and use the CEL bits for it right away (similar to how opting into alpha features lets you do other things that aren't n-1 compatible right away, like validation ratcheting).

jpbetz replied:
> StoredExpressions assumes gated libraries are always available and only goes off of IntroducedVersion / RemovedVersion, since stuff in storage could have gotten in when a feature gate was enabled.

Perfect.

> Making it override IntroducedVersion / RemovedVersion for NewExpressions allows you to opt into an alpha feature and use the CEL bits for it right away (similar to how opting into alpha features lets you do other things that aren't n-1 compatible right away, like validation ratcheting).

I like how this solves both for feature gating and for immediate use of CEL libraries for alpha features (DRA has also needed this).

pohly replied:
Does that mean that I can remove the `IntroducedVersion: version.MajorMinor(1, 0),` hack once this lands and I do... something else instead? :sweat_smile:

jpbetz replied:
Yep, this will cover the DRA use case (using a new CEL library immediately when introducing alpha features) as well as the need to support feature gates.  @pohly can you leave a link to the DRA code so I can be sure to follow up?

liggitt commented:
uh... that's weird
`pull-kubernetes-typecheck — Job succeeded.`

yet:
```
ERROR: staging/src/k8s.io/dynamic-resource-allocation/structured/namedresources/cel/compile.go:1: : # k8s.io/dynamic-resource-allocation/structured/namedresources/cel [k8s.io/dynamic-resource-allocation/structured/namedresources/cel.test]
ERROR: staging/src/k8s.io/dynamic-resource-allocation/structured/namedresources/cel/compile_test.go:127:14: undefined: Compiler (typecheck)
```

liggitt commented:
/retest

liggitt commented:
/retest

jpbetz commented on staging/src/k8s.io/dynamic-resource-allocation/structured/namedresources/cel/compile.go:
```diff
@@ -38,9 +39,17 @@ const (
+	lazyCompiler     *compiler
 )
 
+func GetCompiler() *compiler {
```
Add comment clarifying that this shouldn't be called from inits?

jpbetz commented:
/lgtm
/approve
For CEL related changes.  We'll follow up on the CEL related clean-up post merge of this and keep a close eye on usages of `GetCompiler()` in the near term.

pacoxu commented:
/test pull-kubernetes-integration

pacoxu commented:
`pull-kubernetes-integration` keeps failing 3 times or the same test `TestEncryptAll/encrypt_all_resources` in a row. 

It seems to be a failing CI, not a flake.

liggitt commented:
/hold

for @enj review, and for me to check the test failure

liggitt commented:
looks like the failure is:
```
    kms_transformation_test.go:668: ValidatingAdmissionPolicy.admissionregistration.k8s.io "vap1a1" is invalid: spec.validations[0].expression: Invalid value: "object.spec.replicas <= params.maxReplicas": compilation failed: ERROR: <input>:1:1: undeclared reference to 'object' (in container '')
         | object.spec.replicas <= params.maxReplicas
         | ^
        ERROR: <input>:1:25: undeclared reference to 'params' (in container '')
         | object.spec.replicas <= params.maxReplicas
         | ........................^
```
which appears to be something wonky with the CEL env in the server run by that test. will dig

liggitt commented:
Looks like picking just the `Move CEL env initialization out of package init()` commit on top of `master` is enough to make this test fail, and discarding either the namedresources changes OR the vap changes from that commit is enough to make it start passing again. Looks like there's some weird side-effecty stuff happening in package `init()` today that making those constructions lazy disrupts... will need to dig more to untangle that bit and will update that commit before merge.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/environment/environment.go:
```diff
@@ -231,13 +239,26 @@ func (e *EnvSet) Extend(options ...VersionedOptions) (*EnvSet, error) {
 	var declTypes []*apiservercel.DeclType
 
 	for _, opt := range opts {
+		var allowedByFeatureGate, allowedByVersion bool
```
```suggestion
		allowedByFeatureGate := true
		allowedByVersion := false
```

The variables were not being added to `NewExpressions` for `mustBuildEnvs` when `opt.FeatureEnabled == nil` which set `allowedByFeatureGate` to `false`.

jpbetz replied:
Actually, nevermind, this isn't quite enough.

liggitt replied:
> The variables were not being added to NewExpressions for mustBuildEnvs when opt.FeatureEnabled == nil (which was causing allowedByFeatureGate to remain false).

but that falls back to IntroducedVersion / RemovedVersion logic, which is what I want... 

also, the thing failing the test is not NewExpressions, the VAP API write and validation persists fine... it's the compilation by the VAP *admission* plugin (which I *hope* is using the stored expressions env) that was failing to resolve the root variables

jpbetz commented:
> Looks like picking just the `Move CEL env initialization out of package init()` commit on top of `master` is enough to make this test fail, and discarding either the namedresources changes OR the vap changes from that commit is enough to make it start passing again. Looks like there's some weird side-effecty stuff happening in package `init()` today that making those constructions lazy disrupts... will need to dig more to untangle that bit and will update that commit before merge.

If you accept this suggestion I think it will pass: https://github.com/kubernetes/kubernetes/pull/125571/files#r1662718383

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/environment/environment.go:
```diff
@@ -231,13 +239,26 @@ func (e *EnvSet) Extend(options ...VersionedOptions) (*EnvSet, error) {
 	var declTypes []*apiservercel.DeclType
 
 	for _, opt := range opts {
+		var allowedByFeatureGate, allowedByVersion bool
+		if opt.FeatureEnabled != nil && honorFeatureGateEnablement {
+			// Feature-gate-enabled libraries must follow compatible default feature enablement.
+			// Enabling alpha features in their first release enables libraries the previous API server is unaware of.
+			allowedByFeatureGate = opt.FeatureEnabled()
+			if !allowedByFeatureGate {
+				continue
+			}
+		}
 		if compatVer.AtLeast(opt.IntroducedVersion) && (opt.RemovedVersion == nil || compatVer.LessThan(opt.RemovedVersion)) {
+			allowedByVersion = true
+		}
+
+		if allowedByFeatureGate || allowedByVersion {
```
```suggestion
		allowedByFeatureGate := true
		allowedByVersion := false
		if opt.FeatureEnabled != nil && honorFeatureGateEnablement {
			// Feature-gate-enabled libraries must follow compatible default feature enablement.
			// Enabling alpha features in their first release enables libraries the previous API server is unaware of.
			allowedByFeatureGate = opt.FeatureEnabled()
		}
		if compatVer.AtLeast(opt.IntroducedVersion) && (opt.RemovedVersion == nil || compatVer.LessThan(opt.RemovedVersion)) {
			allowedByVersion = true
		}

		if allowedByFeatureGate && allowedByVersion {
```

I think this is what was intended.

jpbetz replied:
OK, this doesn't help :(  The logic is equivalent so the problem is the same.

The problem is that environment extension is dropping variable additions due to:

https://github.com/kubernetes/kubernetes/blob/403301bfdf2c7312591077827abd2e72f445a53a/test/integration/controlplane/transformation/transformation_test.go#L119

(`compatVer.AtLeast(opt.IntroducedVersion` is false).

Setting `BinaryVersion: "1.0"` would seem like the obvious fix.  @siyuanfoundation are you aware of any problems this would cause?

jpbetz replied:
@siyuanfoundation pointed out to me offline that the 0.0 binary version is to needed in this particular set of test to make sure ALL versions are covered.

Instead we can set our "CEL epoch version" to 0.0 everywhere. E.g.: https://github.com/kubernetes/kubernetes/pull/125848

It works locally.

liggitt replied:
ah, fascinating ... so... this test *should* have been failing in master, but the init()-called env initialization snuck in libraries before the test-set binary version pushed it back to 0.0 which *should* have made *all* the libraries declaring their available version as 1.0 unavailable.

> @siyuanfoundation pointed out to me offline that the 0.0 binary version is to needed in this particular set of test to make sure ALL versions are covered.

Given this test is specifically testing kube-apiserver versions and transformation which didn't exist prior to 1.0, I don't think we care about prior to 1.0.

> Instead we can set our "CEL epoch version" to 0.0 everywhere. E.g.: https://github.com/kubernetes/kubernetes/pull/125848

I'd like to minimize the delta for existing unrelated code in this PR... I'd propose fixing the test to `BinaryVersion: "1.0"` and following up if we want to change the epoch to 1.0. I think a better future pattern would be to tie libraries to feature gates rather than artificial epochs earlier than the version when they were really introduced.

liggitt replied:
actually, I don't understand how this test is expecting to work with BinaryVersion 0.0 OR 1.0... the APIs required to actually start a kube-apiserver don't exist at those versions

liggitt replied:
commented on the PR that added this at https://github.com/kubernetes/kubernetes/pull/122891/files#r1663277481

liggitt replied:
https://github.com/kubernetes/kubernetes/pull/125848 cleaned up that special-case use (thanks), rebased this after that PR merged

BenTheElder commented:
/triage accepted

liggitt commented:
/retest

carlory commented on plugin/pkg/auth/authorizer/node/node_authorizer.go:
```diff
@@ -138,8 +142,15 @@ func (r *NodeAuthorizer) Authorize(ctx context.Context, attrs authorizer.Attribu
 		case resourceSlice:
 			return r.authorizeResourceSlice(nodeName, attrs)
+		case nodeResource:
+			if r.features.Enabled(features.AuthorizeNodeWithSelectors) {
```
It requires AuthorizeWithSelectors to be enabled. Should we also check this feature-gate in this line?

liggitt replied:
validation at startup ensures AuthorizeNodeWithSelectors cannot be enabled if AuthorizeWithSelectors is disabled

liggitt commented:
@pohly can you fix or remove the pull-kubernetes-kind-classic-dra job?

liggitt committed:
Add integration test exercising webhook selector authz

liggitt commented:
Added an integration test exercising selectors in normal API requests and in SAR requests, using them in authz webhook matchConditions and observing them in an authz webhook.

That revealed a couple issues:
1. SAR → AuthzAttributes translation dropped _all_ requirements if a single requirement was not understood, instead of just dropping the invalid requirements and propagating the rest
2. SAR → AuthzAttributes translation was sorting requirements and values, which meant SAR → AuthzAttributes → SAR did not end up sending the requirements and values to the backend webhooks in the same order they were submitted, which was weird

Delta to fixup those issues and add the test in https://github.com/kubernetes/kubernetes/compare/18ccea7d45b703728382190962cebfa9ea8ae394..b380e678881139a564d66410fd8486696c4ffc2e

Ready for final review, then I'll squash my fixup commits for David's commits in.

pohly commented:
> @pohly can you fix or remove the pull-kubernetes-kind-classic-dra job?

I need it to test the new feature gate in https://github.com/kubernetes/kubernetes/pull/125488. Please ignore it for a few more days (fingers crossed...).

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/request/requestinfo.go:
```diff
@@ -238,11 +249,26 @@ func (r *RequestInfoFactory) NewRequestInfo(req *http.Request) (*RequestInfo, er
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(genericfeatures.AuthorizeWithSelectors) {
+		if !verbViaPathPrefix && (requestInfo.Verb == "list" || requestInfo.Verb == "watch" || requestInfo.Verb == "deletecollection") {
```
`verbViaPathPrefix` is an interesting exclusion.  I guess I can't think of a special verb case where I'd want the selectors

liggitt replied:
watch and proxy were the only two special verbs, and the single-item watch endpoint that doesn't support selectors really threw things for a loop... I excluded it because I didn't feel bad telling anyone who wants the benefit of narrower authz to use the non-deprecated watch endpoint

deads2k commented on staging/src/k8s.io/apiserver/pkg/cel/library/cost.go:
```diff
@@ -106,7 +152,7 @@ func (l *CostEstimator) CallCost(function, overloadId string, args []ref.Val, re
 			return &cost
 		}
-	case "masked", "prefixLength", "family", "isUnspecified", "isLoopback", "isLinkLocalMulticast", "isLinkLocalUnicast":
+	case "masked", "prefixLength", "family", "isUnspecified", "isLoopback", "isLinkLocalMulticast", "isLinkLocalUnicast", "isGlobalUnicast":
```
`isGlobalUnicast` was just missing before?

liggitt replied:
yeah, got caught by the test-only fail-on-missing-cost-estimate change I made

we got lucky that the fallback "this costs 1" was actually reasonable for this setter

deads2k commented on staging/src/k8s.io/apiserver/pkg/cel/environment/environment.go:
```diff
@@ -175,7 +175,15 @@ type VersionedOptions struct {
+	// FeatureEnabled returns true if these options are enabled by feature gates,
+	// and returns false if these options are not enabled due to feature gates.
+	//
+	// This takes priority over IntroducedVersion / RemovedVersion for the NewExpressions environment.
```
takes priority in both directions?  If IntroducedVersion is after the desired compatible version, if the featuregate is enabled, the VersionedOption is enabled?

liggitt replied:
Correct. This lets someone opt into an alpha feature in its first release and use a CEL feature unknown to the previous release, just like opting into an alpha feature will let you use API fields unknown to the previous release.

deads2k commented on staging/src/k8s.io/apiserver/pkg/authorization/cel/compile.go:
```diff
@@ -20,22 +20,32 @@ import (
 	Program            cel.Program
 	ExpressionAccessor ExpressionAccessor
+
+	UsesFieldSelector bool
+	UsesLabelSelector bool
```
I'm not entirely clear on why we need this.  Doesn't the parse select the features based on the version of CEL compatibility requested?

liggitt replied:
it's much more complex/expensive to transform the selectors into the unstructured object the matchConditions operate on than all the other request attributes, so if the expressions aren't using those fields, I skip populating the variables for CEL to evaluate on

liggitt replied:
I'll add a comment explaining this

deads2k commented on pkg/features/kube_features.go:
```diff
@@ -67,6 +67,14 @@ const (
+	//
+	// Make the Node authorizer use fine-grained selector authorization.
+	// Requires AuthorizeWithSelectors to be enabled.
+	AuthorizeNodeWithSelectors featuregate.Feature = "AuthorizeNodeWithSelectors"
```
Huh.  I'm surprised by the addition.  Concerned about clients using kubelet creds to read pods in general?

liggitt replied:
Yeah... we know people are piggy backing on kubelet creds, so it seems possible. I expect to graduate them together, but this tightening seemed logically separate from the "send more info to webhooks" bit and more likely to need a one-release escape hatch to me.

deads2k commented on plugin/pkg/auth/authorizer/node/node_authorizer.go:
```diff
@@ -335,6 +346,79 @@ func (r *NodeAuthorizer) authorizeResourceSlice(nodeName string, attrs authorize
+	case "":
+		switch attrs.GetVerb() {
+		case "create", "update", "patch":
+			// Use the NodeRestriction admission plugin to limit a node to creating/updating its own API object.
```
prior to this change, when node/A attempted to update node/B, didn't this authorizer return false because there was no path from node/A to node/B?  Why change this here for update/patch?

liggitt replied:
> prior to this change, when node/A attempted to update node/B, didn't this authorizer return false because there was no path from node/A to node/B?

No, prior to this change, the node authorizer allowed Node updates unconditionally. Node and Pod resources weren't even in this switch statement:

https://github.com/kubernetes/kubernetes/blob/5fc7032a0e509d41890573d5ea2beec880593bda/plugin/pkg/auth/authorizer/node/node_authorizer.go#L114-L143

so all requests for those resources fell through to the generic RBAC rules for nodes and pods:

https://github.com/kubernetes/kubernetes/blob/f2428d66cc2256fcbe2a486b1403603530f46c6f/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L119-L123

https://github.com/kubernetes/kubernetes/blob/f2428d66cc2256fcbe2a486b1403603530f46c6f/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L128-L139

We completely relied on NodeRestriction admission to guard writes, and just broadly allowed reads.

> Why change this here for update/patch?

This PR translates those permissions from RBAC rules to coded rules, and locks down the reads. It's not changing the existing write permissions. We still rely on NodeRestriction to guard writes.

I added positive and negative unit tests for these resources, adjusted the node authorizer unit tests to run with this feature both enabled and disabled, and ensure the results are identical (except the reads of unrelated objects which are now more locked down).

deads2k commented on plugin/pkg/auth/authorizer/node/node_authorizer.go:
```diff
@@ -335,6 +346,79 @@ func (r *NodeAuthorizer) authorizeResourceSlice(nodeName string, attrs authorize
+			// allow a scoped fieldSelector
+			reqs, _ := attrs.GetFieldSelector()
+			for _, req := range reqs {
+				if req.Field == "spec.nodeName" && req.Operator == selection.Equals && req.Value == nodeName {
+					return authorizer.DecisionAllow, "", nil
+				}
```
we always use equals, not in?

liggitt replied:
this is the actual effective parsed field selector, which only supports Equals and NotEquals:

https://github.com/kubernetes/kubernetes/blob/f2428d66cc2256fcbe2a486b1403603530f46c6f/staging/src/k8s.io/apimachinery/pkg/fields/selector.go#L447-L452

https://github.com/kubernetes/kubernetes/blob/f2428d66cc2256fcbe2a486b1403603530f46c6f/staging/src/k8s.io/apimachinery/pkg/fields/selector.go#L110-L113

https://github.com/kubernetes/kubernetes/blob/f2428d66cc2256fcbe2a486b1403603530f46c6f/staging/src/k8s.io/apimachinery/pkg/fields/selector.go#L158-L161

In/NotIn/Exists/DoesNotExist are not a thing yet for the internal parsed field selector.

liggitt commented:
/retest

liggitt commented:
/skip

deads2k commented:
lgtm

please squash in the fixups and we can merge.

enj reviewed (commented):
Minor comments.

liggitt commented on staging/src/k8s.io/apiserver/pkg/authorization/cel/compile.go:
```diff
@@ -96,18 +106,53 @@ func (c compiler) CompileCELExpression(expressionAccessor ExpressionAccessor) (C
+				return
+			}
+			args := e.AsCall().Args()
+			if len(args) != 2 || args[1].Kind() != celast.LiteralKind || args[1].AsLiteral().Type() != cel.StringType {
```
args[0] is the receiver (the authorizer instance we're chaining)

liggitt replied:
added a comment

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/request/requestinfo.go:
```diff
@@ -151,13 +160,15 @@ func (r *RequestInfoFactory) NewRequestInfo(req *http.Request) (*RequestInfo, er
 	currentParts = currentParts[1:]
 
 	// handle input of form /{specialVerb}/*
+	verbViaPathPrefix := false
```
path-prefixed watch supports both the multi-object version which supports selectors:

`/api/v1/watch/namespaces/myns/pods`

and an individual object version which does not:

`/api/v1/watch/namespaces/myns/pods/mypod`

it got gorpy to distinguish between the two so I didn't bother and just limited the selector authorization to the non-deprecated watch endpoints

liggitt replied:
added a comment explaining why

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/request/requestinfo.go:
```diff
@@ -238,11 +249,26 @@ func (r *RequestInfoFactory) NewRequestInfo(req *http.Request) (*RequestInfo, er
+			// interestingly these are parsed above, but the current structure there means that if one (or anything) in the
+			// listOptions fails to decode, the field and label selectors are lost.
+			// therefore, do the straight query param read here.
+			if vals := req.URL.Query()["fieldSelector"]; len(vals) > 0 {
```
yup, see `"selector on repeated param matches parsed param"`

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/request/requestinfo.go:
```diff
@@ -238,11 +249,26 @@ func (r *RequestInfoFactory) NewRequestInfo(req *http.Request) (*RequestInfo, er
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(genericfeatures.AuthorizeWithSelectors) {
+		if !verbViaPathPrefix && (requestInfo.Verb == "list" || requestInfo.Verb == "watch" || requestInfo.Verb == "deletecollection") {
```
done

liggitt commented on staging/src/k8s.io/apiserver/pkg/authorization/cel/compile.go:
```diff
@@ -96,18 +106,53 @@ func (c compiler) CompileCELExpression(expressionAccessor ExpressionAccessor) (C
+
+	var usesFieldSelector, usesLabelSelector bool
+	celast.PreOrderVisit(celast.NavigateAST(celAST), celast.NewExprVisitor(func(e celast.Expr) {
+		var fieldName string
```
done

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/request/requestinfo_test.go:
```diff
@@ -261,5 +322,8 @@ func TestFieldSelectorParsing(t *testing.T) {
 		if e, a := tc.expectedVerb, apiRequestInfo.Verb; e != a {
 			t.Errorf("%s: expected verb %v, actual %v", tc.name, e, a)
 		}
+		if e, a := tc.expectedFieldSelector, apiRequestInfo.FieldSelector; e != a {
```
added

liggitt commented:
rebased and squashed, @enj's comments addressed in https://github.com/kubernetes/kubernetes/compare/36dac5dabfc5542da481ffa61df85d8a7ed72858..5f22dd7c1a455d00c1650d2ce2de4f4110591030

enj commented:
/lgtm

liggitt commented:
/skip

liggitt commented:
/hold cancel

ritazh commented on plugin/pkg/auth/authorizer/node/node_authorizer.go:
```diff
@@ -323,22 +334,104 @@ func (r *NodeAuthorizer) authorizeResourceSlice(nodeName string, attrs authorize
+			// only allow a scoped fieldSelector
+			reqs, _ := attrs.GetFieldSelector()
+			for _, req := range reqs {
+				if req.Field == "nodeName" && req.Operator == selection.Equals && req.Value == nodeName {
```
only the equal operator works here? do we validate the requirements to ensure its equals and return errors or errors when its not a supported operator?

liggitt replied:
This is the internal in-memory authorization attributes representation, which only supports Equals/NotEquals today. 

in SubjectAccessReview structured API serialized form, we use In/NotIn, and we do validate when translating to/from authorization attributes.

ritazh replied:
Do you mean https://github.com/kubernetes/kubernetes/pull/125571/files#diff-3e593550a88b0db529a9b865288964c112d96b6807e58fee301289206fb85398R145

liggitt replied:
there are two paths to get authorization attributes:
1. SubjectAccessReview → authorization attributes translator - that's the helper method you linked to
2. normal `fieldSelector` query parameter parsing of actual list / watch API requests - that produces `hasTerm` instances which only support the Equals operator - https://github.com/kubernetes/kubernetes/blob/a8d354bf398a0f962179f4a8d1c4c8a70347190c/staging/src/k8s.io/apimachinery/pkg/fields/selector.go#L110-L116

pohly commented:
When I try to bring up a cluster with the selector features enabled, kubelet fails:
```console
$ FEATURE_GATES=AuthorizeWithSelectors=true,AuthorizeNodeWithSelectors=true KUBELET_RESOLV_CONF="/etc/resolv-9999.conf" DNS_ADDON="coredns" CGROUP_DRIVER=systemd CONTAINER_RUNTIME_ENDPOINT=unix:///var/run/crio/crio.sock LOG_LEVEL=6 ENABLE_CSI_SNAPSHOTTER=false API_SECURE_PORT=6444 ALLOW_PRIVILEGED=1 ./hack/local-up-cluster.sh -O
...
No resources found
127.0.0.1   NotReady   <none>   2s    v1.31.0-beta.0.189+815efa2baae645-dirty
error: timed out waiting for the condition on nodes/127.0.0.1

$ $ tail -10 /tmp/kubelet.log 
I0721 18:01:04.697655  696243 config.go:104] "Looking for sources, have seen" sources=["api","file"] seenSources={"file":{}}
E0721 18:01:04.765706  696243 kubelet_node_status.go:453] "Error getting the current node from lister" err="node \"127.0.0.1\" not found"
I0721 18:01:04.797139  696243 config.go:104] "Looking for sources, have seen" sources=["api","file"] seenSources={"file":{}}
E0721 18:01:04.866719  696243 kubelet_node_status.go:453] "Error getting the current node from lister" err="node \"127.0.0.1\" not found"
I0721 18:01:04.897142  696243 config.go:104] "Looking for sources, have seen" sources=["api","file"] seenSources={"file":{}}
I0721 18:01:04.919951  696243 round_trippers.go:553] GET https://localhost:6444/apis/storage.k8s.io/v1/csinodes/127.0.0.1 404 Not Found in 2 milliseconds
I0721 18:01:04.922170  696243 round_trippers.go:553] GET https://localhost:6444/api/v1/nodes/127.0.0.1 403 Forbidden in 2 milliseconds
I0721 18:01:04.922256  696243 nodeinfomanager.go:401] Failed to publish CSINode: nodes "127.0.0.1" is forbidden: User "system:node:127.0.0.1" cannot get resource "nodes" in API group "" at the cluster scope: no relationship found between node '127.0.0.1' and this object
E0721 18:01:04.922281  696243 csi_plugin.go:305] Failed to initialize CSINode: error updating CSINode annotation: timed out waiting for the condition; caused by: nodes "127.0.0.1" is forbidden: User "system:node:127.0.0.1" cannot get resource "nodes" in API group "" at the cluster scope: no relationship found between node '127.0.0.1' and this object
F0721 18:01:04.922294  696243 csi_plugin.go:319] Failed to initialize CSINode after retrying: timed out waiting for the condition
```

It works when `FEATURE_GATES` is empty.

Some of these parameters are specific to my machine, I'm just mentioning them for the sake of completeness.

However, https://testgrid.k8s.io/sig-testing-kind#kind-master-alpha is green?

liggitt commented:
Huh… I'll look asap. I suppose we could switch the node `get` check to just use the object name instead of the graph, but I'm puzzled why the unit tests for node gets are working if the graph check wasn't working properly

liggitt commented on plugin/pkg/auth/authorizer/node/node_authorizer.go:
```diff
@@ -323,22 +334,104 @@ func (r *NodeAuthorizer) authorizeResourceSlice(nodeName string, attrs authorize
+			// Use the NodeRestriction admission plugin to limit a node to creating/updating its own API object.
+			return authorizer.DecisionAllow, "", nil
+		case "get", "list", "watch":
+			return r.authorize(nodeName, nodeVertexType, attrs)
```
this could (and probably just should) check the request name in the same as the node name, rather than using the graph

liggitt commented:
> I'm puzzled why the unit tests for node gets are working if the graph check wasn't working properly

oh, I think it's ordering / chicken-and-egg… the unit test creates all the object then makes sure the graph authorizes properly, but I bet the real kubelet has a code path that tries to check if the node object exists before it creates it. Switch just the Node read authz from using the graph to simply comparing the name will resolve that issue.

sftim commented:
Changelog suggestion
```diff
-kube-apiserver: adds an alpha AuthorizeWithSelectors feature that includes field and label selector information from requests in webhook authorization calls; adds an alpha AuthorizeNodeWithSelectors feature that makes the node authorizer limit requests from node API clients to get / list / watch its own Node API object, and to get / list / watch its own Pod API objects. Clients using kubelet credentials to read other nodes or unrelated pods must change their authentication credentials (recommended), adjust their usage, or grant broader read access independent of the node authorizer.
+Added alpha features to allow API server authz to check the context of requests:
+- The `AuthorizeWithSelectors` feature gate enables including field and label selector information
+  from requests in webhook authorization calls.
+- The `AuthorizeNodeWithSelectors` feature gate changes node authorizer behavior to limit requests from node
+  API clients, so that each Node can only get / list / watch its own Node API object, and can also only
+  get / list / watch Pod API objects bound to that node.
+  Clients using kubelet credentials to read other nodes or unrelated pods must change their authentication
+  credentials (recommended), adjust their usage, or obtain broader read access independent of the node authorizer.
```
